File: DataGridView ComboBox allow navigation in drop down/[C#][VB]-DataGridView ComboBox allow navigation in drop down/C# and VB/CustomerDataGridView_CS/MyDataGridView.cs
using System.Windows.Forms;

namespace CustomerDataGridView_CS
{
    public class MyDataGridView : DataGridView
    {
        protected override bool ProcessDataGridViewKey(KeyEventArgs e)
        {
            if (e.KeyData == Keys.Left || e.KeyData == Keys.Right)
            {
                if (this.EditingControl != null)
                {
                    if (this.EditingControl.GetType() ==
                    typeof(DataGridViewComboBoxEditingControl))
                    {
                        ComboBox control = this.EditingControl as ComboBox;
                        if (control.DropDownStyle !=
                        ComboBoxStyle.DropDownList)
                        {
                            switch (e.KeyData)
                            {
                                case Keys.Left:
                                    if (control.SelectionStart > 0)
                                    {
                                        control.SelectionStart--;
                                        return true;
                                    }
                                    break;

                                case Keys.Right:
                                    if (control.SelectionStart <
                                    control.Text.Length)
                                    {
                                        control.SelectionStart++;
                                        return true;
                                    }
                                    break;
                            }
                        }
                    }
                }
            }

            return base.ProcessDataGridViewKey(e);
        }
    }
}

File: DataGridView ComboBox allow navigation in drop down/[C#][VB]-DataGridView ComboBox allow navigation in drop down/C# and VB/CustomerDataGridView_VB/MyDataGridView.vb
Public Class MyDataGridView
    Inherits DataGridView

    Protected Overrides Function ProcessDataGridViewKey(ByVal e As KeyEventArgs) As Boolean

        If e.KeyData = Keys.Left OrElse e.KeyData = Keys.Right Then
            If Me.EditingControl IsNot Nothing Then
                If Me.EditingControl.GetType() Is GetType(DataGridViewComboBoxEditingControl) Then
                    Dim control As ComboBox = TryCast(Me.EditingControl, ComboBox)
                    If control.DropDownStyle <> ComboBoxStyle.DropDownList Then
                        Select Case e.KeyData
                            Case Keys.Left
                                If control.SelectionStart > 0 Then
                                    control.SelectionStart -= 1
                                    Return True
                                End If
                            Case Keys.Right
                                If control.SelectionStart < control.Text.Length Then
                                    control.SelectionStart += 1
                                    Return True
                                End If
                        End Select
                    End If
                End If
            End If

        End If

        Return MyBase.ProcessDataGridViewKey(e)

    End Function
End Class